File: adt/set/sparse-set.pl
#!/usr/bin/env perl -w
#
# ===============================================================
#
# Filename:	set.pl
#
# Author:		Oxnz
# Email:		yunxinyi@gmail.com
# Created:		[2014-11-23 04:34:43 CST]
# Last-update:	2014-11-23 04:34:43 CST
# Description: ANCHOR
#
# Version:		0.0.1
# Revision:	[None]
# Revision history:	[None]
# Date Author Remarks:	[None]
# ref: http://programmingpraxis.com/2012/03/09/sparse-sets/
# License:
# Copyright (c) 2013 Oxnz
#
# Distributed under terms of the [LICENSE] license.
# [license]
#
# ===============================================================
#

use strict;
use warnings;

package SparseSet;

use constant { TRUE => 1, FALSE => 0 };

sub new {
	my $class = shift;
	my $self = bless {
		n => 0,
		dense => [],
		sparse => [],
	}, $class;
	return $self;
}

sub clear {
	my $self = shift;
	$self->{n} = 0;
}

sub contain {
	my $self = shift;
	my $val = shift;
	return undef if $self->{n} <= 0;
	my $idx = $self->{sparse}->[$val];
	return undef if not defined($idx) or $idx >= $self->{n};
	defined($idx) or return undef;
	return $val == $self->{dense}->[$idx];
	return 1;
}

sub insert {
	my $self = shift;
	my $val = shift;
	return undef if $self->contain($val);
	$self->{dense}->[$self->{n}] = $val;
	$self->{sparse}->[$val] = $self->{n};
	++$self->{n}
}

sub remove {
	my $self = shift;
	my $val = shift;
	return undef if not $self->contain($val);
	--$self->{n};
	my $topval = $self->{dense}->[$self->{n}];
	return 1 if $val == $topval;
	my $idx = $self->{sparse}->[$val];
	$self->{sparse}->[$topval] = $idx;
	$self->{dense}->[$idx] = $topval;
	return 1;
}

sub inspect {
	my $self = shift;
	print "sparse set: $self->{n}\n";
	if ($self->{n}) {
		print "elems:\n\t";
		for (my $i = 0; $i < $self->{n}; ++$i) {
			print "[$self->{dense}->[$i]]";
		}
		print "\ndense array:\n\t";
		foreach (@{$self->{dense}}) {
			if (defined($_)) {
				print "[$_]";
			} else {
				print "[ ]";
			}
		}
		print "\nsparse array:\n\t";
		foreach (@{$self->{sparse}}) {
			if (defined($_)) {
				print "[$_]";
			} else {
				print "[ ]";
			}
		}
		print "\n";
	}
}

sub DESTROY {
	#print "destroy";
}

package main;

print "test sparse set:\n";
my $set = SparseSet->new(10);
print "new set:\n";
$set->inspect();
foreach (9, 8, 5, 6, 3) {
	$set->insert($_);
}
print "after insert 9 8 5 6 3\n";
$set->inspect();
$set->remove(5);
print "after remove(5):\n";
$set->inspect();
print "after insert 5 7 1 9 0:\n";
foreach (5, 7, 1, 9, 0) {
	$set->insert($_);
}
$set->inspect();

File: adt/sparse_set.cpp
/*
 * Filename:	sparse-dense_unittest.cc
 *
 * Author:		Oxnz
 * Email:		yunxinyi@gmail.com
 * Created:		[2014-11-30 01:53:46 CST]
 * Last-update:	2014-11-30 01:53:46 CST
 * Description: anchor
 *
 * Version:		0.0.1
 * Revision:	[NONE]
 * Revision history:	[NONE]
 * Date Author Remarks:	[NONE]
 *
 * License:
 * Copyright (c) 2013 Oxnz
 *
 * Distributed under terms of the [LICENSE] license.
 * [license]
 *
 */

#include <iostream>
#include <gtest/gtest.h>
#include <unordered_set>
#include "sparse_set.h"

using namespace std;

namespace {
    class SparseSetTest : public ::testing::Test {
	protected:
        SparseSetTest() : s(10) {}
        ~SparseSetTest() {}

		static void SetUpTestCase() {
		}

		static void TearDownTestCase() {
		}

		void SetUp() {
		}

		void TearDown() {
		}
        m::sparse_set s;
	};
}

TEST_F(SparseSetTest, insert) {
	EXPECT_TRUE(s.insert(1));
	EXPECT_EQ(1, s.size());
	EXPECT_FALSE(s.empty());

	EXPECT_FALSE(s.insert(1));
	EXPECT_EQ(1, s.size());
}

TEST_F(SparseSetTest, remove) {
	EXPECT_TRUE(s.insert(3));
	EXPECT_FALSE(s.remove(1));
	EXPECT_TRUE(s.remove(3));
	EXPECT_FALSE(s.remove(3));
	EXPECT_FALSE(s.contain(4));
	EXPECT_TRUE(s.insert(4));
	EXPECT_TRUE(s.contain(4));
}

TEST(sparse_set, empty) {
    size_t n = 10;
    m::sparse_set s(n);
    EXPECT_EQ(s.capacity(), n);
    EXPECT_EQ(0, s.size());
    EXPECT_TRUE(s.empty());
}

TEST(sparse_set, for_each) {
    std::unordered_set<int> sep = {8, 9, 3, 2};
    m::sparse_set s(10);
    for (auto i : sep) s.insert(i);
    std::unordered_set<int> vals;
    s.foreach([&vals](int n)->void * { vals.insert(n); return 0; });
    EXPECT_EQ(sep, vals);
}

TEST(sparse_set, io) {
	std::unordered_set<int> sep = {8, 9, 3, 2};
    m::sparse_set s(10);
	for (int n : sep) {
        EXPECT_FALSE(s.contain(n));
        EXPECT_TRUE(s.insert(n));
        EXPECT_TRUE(s.contain(n));
	}
    for (int i = 0; i < 10; ++i) {
        if (sep.count(i)) EXPECT_TRUE(s.contain(i));
        else EXPECT_FALSE(s.contain(i));
    }
}

TEST(sparse_set, op) {
    std::unordered_set<int> sep = {8, 9, 3, 2};
    m::sparse_set s(10);
    for (auto i : sep) s.insert(i);
    s.remove(2);
    EXPECT_FALSE(s.contain(2));
    s.remove(3);
    EXPECT_FALSE(s.contain(3));
	s.clear();
    EXPECT_FALSE(s.contain(8));
    s.insert(3);
    EXPECT_TRUE(s.contain(3));
	s.insert(7);
    EXPECT_TRUE(s.contain(7));
	s.insert(4);
    EXPECT_TRUE(s.contain(4));
}
